### PR TITLE
feat: add theme provider component documentation

### PR DESCRIPTION
**Pull request by jzempel**
## Description

New pages appear under `Foundations > Theming` nav:

- Palette
- Theme object
- Theme provider
- Utilities

## Detail

Reviewers (@vsrixyz, @rossmoody, @allisonacs) please note that all sections, including API, were documented in this PR rather than auto-generated. As such, it will be good to put :eyes: on all aspects – API sections included. I also made some game-time decisions apart from the previously drafted documentation. Mostly having to do with the segments we were feeling iffy about.

## Checklist

- [x] :ok_hand: website updates are Garden Designer approved (add the designer as a reviewer)
- [x] :black_nib: copy updates are approved (add the content strategist as a reviewer)
- [x] :link: considered opportunities for adding cross-reference URLs (grep for keywords)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

File: src/examples/utilities/components.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    }
+  },
+  {
+    name: 'withTheme',
+    props: {
+      Component: {
+        required: true,
+        type: 'Object',
+        description:
+          'Takes a component to be wrapped by a higher-order component with an added `theme` prop'
```
Comment (vsrixyz) [inline src/examples/utilities/components.ts:166]: This one is kind of hard to parse. What does it mean to "take a component"?

File: src/pages/components/palette.mdx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      <li>
+        <Markdown>
+          To style custom components with Garden, access the palette through the
+          [`theme.palette`](/components/theme-object#palette) API or
+          [`withTheme`](/components/utilities#withTheme) utility instead
+        </Markdown>
+      </li>
+      <li>
+        <Markdown>
+          To extend Garden with reusable components that respond to theming, use the
```
Comment (vsrixyz) [inline src/pages/components/palette.mdx:30]: I think you can lose "together":

To extend Garden with reusable components that respond to theming, use the [`getColor`](/components/utilities#getColor) utility ~together~ with the [`theme.colors`](/components/theme-object#colors) API instead

File: src/pages/components/theme-object.mdx [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+values are weights used by Garden components.
+
+### iconSizes
+
+The icon sizes object corresponds with the small, medium, and large icons
+provided by Garden's [SVG Icon](/design/icons/installation) library.
+
+### lineHeights
+
+Line heights are used in conjunction with [`fontSizes`](#fontSizes) to define
```
Comment (vsrixyz) [inline src/pages/components/theme-object.mdx:188]: You can make this simpler:

Line heights are used ~n conjunction~ with

File: src/pages/components/utilities.mdx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  - theming/src/utils/arrowStyles.ts
+  - theming/src/utils/getColor.ts
+  - theming/src/utils/getLineHeight.ts
+  - theming/src/utils/mediaQuery.ts
+  - theming/src/utils/menuStyles.ts
+  - theming/src/utils/retrieveComponentStyles.ts
+  - theming/src/utils/withTheme.ts
+---
+
+## How to use it
```
Comment (vsrixyz) [inline src/pages/components/utilities.mdx:17]: Should we lose this heading since there is nothing underneath it?

File: src/pages/components/utilities.mdx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+### Menu and arrow styles
+
+Use menu and arrow styling utilities together to ease the production of
```
Comment (vsrixyz) [inline src/pages/components/utilities.mdx:21]: What does "ease the production" mean? Do we mean: Use menu and arrow styling utilities to make it easier to build Garden-styled menu components?

**Review by vsrixyz — COMMENTED** [review]
Only a few suggestions

File: src/examples/utilities/components.ts (hunk 1 above)
Comment (jzempel) [reply src/examples/utilities/components.ts:166]: @vsrixyz I agree. Yet this is parlance that is familiar in React land. Check out https://reactjs.org/docs/higher-order-components.html

File: src/pages/components/utilities.mdx (hunk 4 above)
Comment (jzempel) [reply src/pages/components/utilities.mdx:17]: No. This is standard throughout the component pages and we must retain a valid H1 -> H2 -> H3 for accessibility.

File: src/pages/components/utilities.mdx (hunk 5 above)
Comment (jzempel) [reply src/pages/components/utilities.mdx:21]: Yes. In fact, we can probably simplify to:

> Use menu and arrow styling utilities to build Garden-styled menu components.

File: src/examples/utilities/components.ts (hunk 1 above)
Comment (jzempel) [reply src/examples/utilities/components.ts:166]: ...although, I think I can do better.

> Gets wrapped by a higher-order component with an added `theme` prop

File: src/pages/components/utilities.mdx (hunk 4 above)
Comment (jzempel) [reply src/pages/components/utilities.mdx:17]: Although, should this be "How to use them" as we are referring to Utilities rather than a single Component?

File: src/pages/components/theme-object.mdx [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+### palette
+
+This object lets you reference, modify, and extend Garden’s entire
+[Palette](/components/palette) (with the exception of `palette.product`
+values, which cannot be customized).
+
+### rtl
+
+The RTL theme object value determines whether components will layout
```
Comment (rossmoody) [inline src/pages/components/theme-object.mdx:200]: `lay out`

**Comment by rossmoody** [comment]
Looking so good. Broader topic to consider: Since there is no visual difference between a `code` snippet and a `code` snippet that is also a link, I wonder if it would be better to go with an anchor styling so links don't go unnoticed.

**Comment by jzempel** [comment]
> Since there is no visual difference between a `code` snippet and a `code` snippet that is also a link, I wonder if it would be better to go with an anchor styling so links don't go unnoticed. 

Yesss. I was hoping you would help me make the call on this. Let's work together to determine a direction for this PR and future `website` content.

File: src/pages/components/theme-provider.mdx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+<CodeExample code={NestingCode}>
+  <Nesting />
+</CodeExample>
+
+### Targeting
+
+The theme's [components](/components/theme-object#components) object can
+override styling for a targeted component type. In this example, Garden
+[Tabs](/component/tabs) are restyled to appear below associated panel
```
Comment (hzhu) [inline src/pages/components/theme-provider.mdx:83]: It looks like this URL is broken because the path should be "/components/tabs".

File: src/pages/components/theme-provider.mdx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+### Default
+
+The typical usage of a theme provided to styled components.
+
+import Default from '../../examples/theme-provider/ThemeProviderDefault.tsx';
+import DefaultCode from '!!raw-loader!../../examples/theme-provider/ThemeProviderDefault.tsx';
+
+<CodeExample code={DefaultCode}>
+  <Default />
+</CodeExample>
```
Comment (hzhu) [inline src/pages/components/theme-provider.mdx:36]: The examples look good on the website but seem like it's rendered differently on CodeSandbox. Anyone else experiencing this?

Screen Shot 2020-11-03 at 12 43 51 PM

File: src/pages/components/utilities.mdx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+are applied must be wrapped in a positioned box (`relative`, `absolute`, or
+`fixed`) that has a `z-index` greater than or equal to zero.
+
+<PropSheet components={COMPONENTS} componentName="arrowStyles" headerName="Parameter" />
+
+### getColor
+
+Get a color for the given hue, shade, and theme. The color is resolved with a
+fallback algorithm that succeeds whether or not the particular shade exists
+within the theme.
```
Comment (allisonacs) [inline src/pages/components/utilities.mdx:52]: This wording is confusing to me as a non-developer. I think what it's trying to say is that `getColor` will provide a color even if the particular shade is not part of the theme. 

Consider being more explicit that the shade doesn't have to be in the theme. I think that's what was confusing to me. Also the word "succeeds." Maybe something like "…the algorithm returns a valid color even if the particular shade doesn't exist in the current theme."

**Review by allisonacs — CHANGES_REQUESTED** [review]
minor content edits

File: src/pages/components/theme-provider.mdx (hunk 8 above)
Comment (jzempel) [reply src/pages/components/theme-provider.mdx:36]: Ah, I think it's because I'm using `css` rather than `style`. I'll get this fixed.

File: src/components/MarkdownProvider/components/Configuration.tsx [hunk 10]
```diff
@@ -76,7 +77,8 @@ export const Configuration: React.FC<{ reactPackage: IPackage; components: IComp
       <StyledListItem>
         <StyledListItemLabel isBold>Deps</StyledListItemLabel>
         <StyledMono>
-          npm install react react-dom prop-types styled-components @zendeskgarden/react-theming
+          npm install react react-dom prop-types styled-components
+          {!isTheming && ' @zendeskgarden/react-theming'}
```
Comment (austingreendev) [inline src/components/MarkdownProvider/components/Configuration.tsx:81]: Rather than a separate prop, could we use `reactPackage.packageName === '@zendeskgarden/react-theming'` to dynamically apply this value?

File: src/pages/components/palette.mdx [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      </li>
+    </ul>
+  </Use>
+  <Misuse>
+    <ul>
+      <li>
+        <Markdown>
+          To style custom components with Garden, access the palette through the
+          [`theme.palette`](/components/theme-object#palette) API or
+          [`withTheme`](/components/utilities#withTheme) utility instead
```
Comment (austingreendev) [inline src/pages/components/palette.mdx:25]: Should we reference the `ThemeContext` object as that's the "new" approach provided by styled-components?

Comment (jzempel) [reply src/pages/components/palette.mdx:25]: Yeah, probably. The link to `withTheme` gets to that detail. But I was almost wondering if we wanted to ditch documenting `withTheme` on the website?

Comment (austingreendev) [reply src/pages/components/palette.mdx:25]: To support class components I don't think the HOC is going anywhere on the styled-components side. So I guess that's a good reason to leave it around at least.

Comment (jzempel) [reply src/pages/components/palette.mdx:25]: I just elected to remove the `withTheme` clause here. In this context it is relevant to reference the theme object's key, less relevant to determine how to access (styled vs. non-styled Component) the key.

**Review by hzhu — APPROVED** [review]

**Review by rossmoody — APPROVED** [review]

**Review by allisonacs — APPROVED** [review]
